### DISCUSSION TITLE
Can cub::DeviceSegmentedSort::SortKeys support user-defined compare functors?

### DISCUSSION THREAD

**Discussion by zlwu92** [NVIDIA/cccl#1061] [CUB]
Let's say I have a device array (use cudaMalloc), int* nodelist[], and another device array (use cudaMalloc), int* weight[]...

I want to use cub::DeviceSegmentedSort::SortKeys to sort nodeslit[], but I want to define a compare functor that need to use the weight[]..., can I do this?

**Comment by gevtushenko** [comment]
Hello @zlwu92! Unfortunately, our segmented sort facilities do not support comparison operators. That's because segmented sort relies on radix sort in the cases of large segments and a small number of segments (radix sort is not comparison based).  

Nonetheless, if your weights are of arithmetic type, you could use `cub::DeviceSegmentedSort::SortPairs` with weight as keys and nodelist as values. In this case, values will be permuted the same way as keys. 

In case you need a non-trivial comparison operator, you could use `cub::DeviceMergeSort::SortPairs` if you embed the segment id into the most significant part of your keys.
